### PR TITLE
Add USES model for speech enhancement in diverse conditions

### PR DESCRIPTION
**Pull request by Emrys365**
## What?

This PR adds a new speech enhancement model [USES](https://arxiv.org/abs/2309.17384), which is capable of handling diverse input conditions with a single model:
  * different sampling frequencies
  * different microphone variations (1 or more channels with different array geometries)
  * different signal lengths (with approximately linear complexity)

Along with the USES model implementation, I also update the related functions that are required to train this new model, including:
  * [espnet2/iterators/chunk_iter_factory.py](https://github.com/espnet/espnet/blob/master/espnet2/iterators/chunk_iter_factory.py): A new argument `default_fs` is added to `ChunkIterFactory` to allow adaptive adjustment of `chunk_length` according to the sampling frequency of the batch. The related argument is also added to [espnet2/tasks/abs_task.py](https://github.com/espnet/espnet/blob/master/espnet2/tasks/abs_task.py).
  * [espnet2/train/preprocessor.py](https://github.com/espnet/espnet/blob/master/espnet2/train/preprocessor.py): Add the sampling frequency information in `EnhPreprocessor`. Also improve some existing functions to more flexibility.
  * [espnet2/enh/encoder/stft_encoder.py](https://github.com/espnet/espnet/blob/master/espnet2/enh/encoder/stft_encoder.py) and [espnet2/enh/decoder/stft_decoder.py](https://github.com/espnet/espnet/blob/master/espnet2/enh/decoder/stft_decoder.py): Add two methods `reset_config` and `reconfig_for_fs` to allow adaptive adjustment of `n_fft`, `win_length`, and `hop_length` according to the sampling frequency of the input signal.
  * [espnet2/enh/loss/criterions/time_domain.py](https://github.com/espnet/espnet/blob/master/espnet2/enh/loss/criterions/time_domain.py): Update `MultiResL1SpecLoss` to support different reduction modes and variance normalization before loss calculation.
  * [espnet2/enh/layers/dptnet.py](https://github.com/espnet/espnet/blob/master/espnet2/enh/layers/dptnet.py): Add a new option ("linear") for the argument `activation`, which will be used in the new model.
  * `espnet2/enh/layers/uses.py` and `espnet2/enh/separator/uses_separator.py`: Implementation of the new model.
  * [espnet2/enh/espnet_model.py](https://github.com/espnet/espnet/blob/master/espnet2/enh/espnet_model.py)
    * Add options for variance normalization (and de-normalization) before (after) model forward.
    * Add category-related processing in the model forward method.
    * Add sampling frequency related processing in the model forward method.
    * Add category-specific logging for training, i.e., losses from batches of different categories are reported with different suffixes.
  * [espnet2/torch_utils/recursive_op.py](https://github.com/espnet/espnet/blob/master/espnet2/torch_utils/recursive_op.py): Modify the way data from different nodes are accumulated in the DDP mode to be compatible with the category-specific logging in `espnet2/enh/espnet_model.py`.
  * [espnet2/bin/enh_inference.py](https://github.com/espnet/espnet/blob/master/espnet2/bin/enh_inference.py): Add category-related processing and variance normalization/de-normalization.

## Why?

This new SE model and the related function updates can increase the flexibility of the current SE framework.

## See also
Pre-trained model is available at [HuggingFace](https://huggingface.co/espnet/Wangyou_Zhang_universal_train_enh_uses_refch0_2mem_raw).

### CODE REVIEW

**Comment by LiChenda** [comment]
I'll review it.

**Comment by sw005320** [comment]
@kohei0209, please review this PR.

**Comment by kohei0209** [comment]
Sure, I'll review this PR.

**Comment by sw005320** [comment]
@kohei0209 and @LiChenda, can you review this PR?

File: espnet2/train/preprocessor.py [hunk 1]
```diff
@@ -258,6 +264,13 @@ def _convolve_rir(self, speech, power, rirs, tgt_fs=None, single_channel=False):
         rir = None
         if rir_path is not None:
             rir, fs = soundfile.read(rir_path, dtype=np.float64, always_2d=True)
+
+            if single_channel:
+                num_ch = rir.shape[1]
+                chs = np.random.permutation(num_ch).tolist()
+                rir = rir[:, chs]
```
Comment (kohei0209) [inline espnet2/train/preprocessor.py:271]: If my understanding is correct, `rir` is just shuffled along channel dim but still multi-channel. It might be better to make `rir` single-channel here since the rir convolution becomes redundant.

File: espnet2/train/preprocessor.py [hunk 2]
```diff
@@ -327,7 +335,9 @@ def _add_noise(
                     if len(noise) != nsamples_:
                         raise RuntimeError(f"Something wrong: {noise_path}")
             if single_channel:
-                noise = noise[:, :1]
+                num_ch = noise.shape[1]
+                chs = np.random.permutation(num_ch).tolist()
+                noise = noise[:, chs]
```
Comment (kohei0209) [inline espnet2/train/preprocessor.py:340]: We can also make `noise` single-channel here.

**Review by kohei0209 — COMMENTED** [review]
I left several comments. Could you check them?

File: espnet2/bin/enh_inference.py
Comment (LiChenda) [inline espnet2/bin/enh_inference.py:1]: It looks like some new code blocks are not covered by the test. Could you also update the unit test?

File: espnet2/enh/espnet_model.py [hunk 3]
```diff
@@ -38,6 +41,10 @@ def __init__(
         mask_type: Optional[str] = None,
         flexible_numspk: bool = False,
         extract_feats_in_collect_stats: bool = False,
+        normalize_variance: bool = False,
+        normalize_variance_per_ch: bool = False,
+        categories: list = [],
+        category_weights: list = [],
     ):
```
Comment (LiChenda) [inline espnet2/enh/espnet_model.py:48]: I feel like we need to add a little more doc string here, given that we have quite a few parameters passed in this init function.

File: espnet2/enh/encoder/stft_encoder.py [hunk 4]
```diff
@@ -65,6 +71,22 @@ def forward(self, input: torch.Tensor, ilens: torch.Tensor):
 
         return spectrum, flens
 
+    def reset_config(self):
+        """Reset the configuration of STFT window and hop lengths."""
+        self.reconfig_for_fs(self.default_fs)
+
+    def reconfig_for_fs(self, fs):
```
Comment (LiChenda) [inline espnet2/enh/encoder/stft_encoder.py:78]: How about making the `reconfig_for_fs` and `reset_config` functions private and call them in the `forward` function? I think that will be much clearer and safer. I feel changing the internal property of the `STFT` module somewhere else and forgetting to reset it may lead to hidden dangers. And just passing the `fs` to the `forward` function will be much simpler.

File: espnet2/enh/decoder/stft_decoder.py [hunk 5]
```diff
@@ -82,6 +83,22 @@ def forward(self, input: ComplexTensor, ilens: torch.Tensor):
 
         return wav, wav_lens
 
+    def reset_config(self):
+        """Reset the configuration of iSTFT window and hop lengths."""
+        self.reconfig_for_fs(self.default_fs)
+
+    def reconfig_for_fs(self, fs):
```
Comment (LiChenda) [inline espnet2/enh/decoder/stft_decoder.py:90]: Please see the same comments in the encoder module.

File: espnet2/enh/espnet_model.py [hunk 6]
```diff
@@ -161,6 +189,28 @@ def forward(
...
+        # sampling frequency information about the batch
+        if "utt2fs" in kwargs:
+            # All samples must have the same sampling rate
+            fs = kwargs["utt2fs"][0].item()
+            assert all([fs == kwargs["utt2fs"][0].item() for fs in kwargs["utt2fs"]])
+
+            # Adaptively adjust the STFT encoder/decoder window/hop sizes
+            if isinstance(self.separator, USESSeparator):
+                if isinstance(self.encoder, STFTEncoder):
+                    self.encoder.reconfig_for_fs(fs)
```
Comment (LiChenda) [inline espnet2/enh/espnet_model.py:201]: I suggest removing this call and passing `fs` to the encoder's `forward` function directly.

File: espnet2/enh/espnet_model.py [hunk 7]
```diff
@@ -187,8 +271,16 @@ def forward(
             speech_ref,
             noise_ref,
             dereverb_speech_ref,
+            category,
             num_spk=num_spk,
         )
+
+        if isinstance(self.separator, USESSeparator):
+            # Reset the STFT encoder/decoder window/hop sizes
+            if isinstance(self.encoder, STFTEncoder):
```
Comment (LiChenda) [inline espnet2/enh/espnet_model.py:280]: ditto

**Review by LiChenda — COMMENTED** [review]
Hi, @Emrys365 , I updated some comments for this PR.

File: espnet2/enh/espnet_model.py (hunk 6 above)
Comment (Emrys365) [reply espnet2/enh/espnet_model.py:201]: Good idea!

File: espnet2/enh/espnet_model.py [hunk 8]
```diff
@@ -74,10 +81,31 @@ def __init__(
...
 
-        # Used in espnet2/tasks/abs_task.py for determining whether or not to do
+        # used in espnet2/tasks/abs_task.py for determining whether or not to do
         # collect_feats during collect stats (stage 5).
         self.extract_feats_in_collect_stats = extract_feats_in_collect_stats
 
+        # normalize signal variance before model forward, and revert it back after
+        self.normalize_variance = normalize_variance
+        # normalize signal variance for each channel instead of the whole signal
+        self.normalize_variance_per_ch = normalize_variance_per_ch
```
[the remark this answers is not in the document]
Comment (Emrys365) [reply espnet2/enh/espnet_model.py:91]: Thanks! I will add it.

File: espnet2/enh/layers/dptnet.py [hunk 9]
```diff
@@ -64,11 +64,16 @@ def __init__(
...
-        )
+        if activation.lower() == "linear":
+            self.feed_forward = nn.Sequential(
+                nn.Dropout(p=dropout), nn.Linear(hdim, input_size)
+            )
+        else:
+            activation = get_activation(activation)
+            self.feed_forward = nn.Sequential(
+                activation, nn.Dropout(p=dropout), nn.Linear(hdim, input_size)
+            )
```
[the remark this answers is not in the document]
Comment (Emrys365) [reply espnet2/enh/layers/dptnet.py:76]: OK. This sounds good.

File: espnet2/train/preprocessor.py (hunk 1 above)
Comment (Emrys365) [reply espnet2/train/preprocessor.py:271]: You're right. I will update the code to only keep one channel.

File: espnet2/enh/espnet_model.py (hunk 3 above)
Comment (Emrys365) [reply espnet2/enh/espnet_model.py:46]: done

File: espnet2/enh/espnet_model.py (hunk 6 above)
Comment (Emrys365) [reply espnet2/enh/espnet_model.py]: done

File: espnet2/enh/espnet_model.py (hunk 7 above)
Comment (Emrys365) [reply espnet2/enh/espnet_model.py]: done

File: espnet2/enh/encoder/stft_encoder.py (hunk 4 above)
Comment (Emrys365) [reply espnet2/enh/encoder/stft_encoder.py]: Good idea!

Comment (Emrys365) [reply espnet2/enh/encoder/stft_encoder.py]: done

**Review by LiChenda — APPROVED** [review]
Thank you @Emrys365 ! Now, it looks good to me!

**Review by sw005320 — APPROVED** [review]
LGTM

**Review by kohei0209 — APPROVED** [review]
@Emrys365  Looks good to me, thank you for reflecting the comments!

**Comment by sw005320** [comment]
Thanks!
